### PR TITLE
Added test to automate https://bugzilla.redhat.com/show_bug.cgi?id=1418412

### PR DESCRIPTION
**Pull request by chbrown13**
Pull request for issue #4864 to automatically test that multibyte and latin1 org names are displayed with the same width.

### CODE REVIEW

**Comment by chbrown13** [comment]
Test results

[root@localhost robottelo-fork]# pytest tests/foreman/cli/test_organization.py -k 'test_positive_multibyte_latin1_org_names'
==================================================================================================================================== test session starts ====================================================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.7, py-1.4.34, pluggy-0.4.0
rootdir: /home/chris/Documents/robottelo-fork, inifile:
plugins: cov-2.5.1, mock-1.6.0, xdist-1.17.1, services-1.2.1
collected 70 items 
2017-06-28 14:39:02 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/cli/test_organization.py .

==================================================================================================================================== 69 tests deselected ====================================================================================================================================
========================================================================================================================= 1 passed, 69 deselected in 137.14 seconds =========================================================================================================================

File: tests/foreman/cli/test_organization.py [hunk 1]
```diff
@@ -1693,3 +1694,36 @@ def test_positive_info_by_name(self):
...
+        """Hammer Multibyte and Latin-1 Org names break list pagination
+
+        :id: 4fa0afe7-6d0a-4c3e-a0fc-4ecb95c50fc9
+
+        :BZ: 1418412
+
+        :expectedresults: Multibyte and latin1 names need to be
+            displayed with consistent spacing
+
+        :CaseImportance: Critical
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_organization.py]: why is this critical?

File: tests/foreman/cli/test_organization.py [hunk 2]
```diff
@@ -1693,3 +1694,36 @@ def test_positive_info_by_name(self):
...
+            make_org({'name': org})
+        org_list = Org.list(output_format='table')
+        for org in org_names:
+            Org.delete({'name': org})
+        if org_list:
+            width = len(org_list[0])
+            for org_str in org_list:
+                if org_str:
+                    n = 0
+                    for c in org_str:
```
Comment (JacobCallahan) [inline tests/foreman/cli/test_organization.py]: this line makes the ```if org_str:``` line unnecessary

File: tests/foreman/cli/test_organization.py (hunk 1 above)
Comment (chbrown13) [inline tests/foreman/cli/test_organization.py]: My bad, copy-paste error

File: tests/foreman/cli/test_organization.py (hunk 2 above)
Comment (chbrown13) [inline tests/foreman/cli/test_organization.py]: I think we need to have a check here because the last item in the list is empty string and will always fail the assertion

File: tests/foreman/cli/test_organization.py [hunk 3]
```diff
@@ -1693,3 +1694,32 @@ def test_positive_info_by_name(self):
         org = make_org()
         result = Org.info({'name': org['name']})
         self.assertEqual(org['id'], result['id'])
+
+    @tier1
+    @skip_if_bug_open('bugzilla', 1418412)
```
Comment (oshtaier) [inline tests/foreman/cli/test_organization.py]: Redundant statement as defect is already in `VERIFIED` state

**Comment by oshtaier** [comment]
Please, make sure that you have one commit in PR. It is not a strict rule, but we try to follow it.
In the other hand, it is fine to have more commits when you addressing other people review comments to make the process easier

File: tests/foreman/cli/test_organization.py [hunk 4]
```diff
@@ -1693,3 +1694,32 @@ def test_positive_info_by_name(self):
...
+        :BZ: 1418412
+
+        :expectedresults: Multibyte and latin1 names need to be
+            displayed with consistent spacing
+        """
+        org_names = valid_org_names_list()
+        for org in org_names:
+            make_org({'name': org})
+        org_list = Org.list(output_format='table').remove('')
+        for org in org_names:
```
Comment (oshtaier) [inline tests/foreman/cli/test_organization.py]: Do we make cleanup procedure here? Or what the purpose to remove all created organizations?

File: tests/foreman/cli/test_organization.py [hunk 5]
```diff
@@ -1693,3 +1694,32 @@ def test_positive_info_by_name(self):
...
+
+        :BZ: 1418412
+
+        :expectedresults: Multibyte and latin1 names need to be
+            displayed with consistent spacing
+        """
+        org_names = valid_org_names_list()
+        for org in org_names:
+            make_org({'name': org})
+        org_list = Org.list(output_format='table').remove('')
```
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: `remove()` returns None, your org_list will always appear empty.

Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: In addition, `remove()` will delete only the first found element (empty string in your case), so in case there're few empty lines, this approach won't work. Try
```python
org_list = [line for line in Org.list(output_format='table') if line]
```
instead.

File: tests/foreman/cli/test_organization.py (hunk 4 above)
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: Same question. Cleanup started too early, in case organization deletion is broken, the test won't even test anything. I'd rather either left orgs untouched or removed them as the last step, in some `finally:` block.

File: tests/foreman/cli/test_organization.py [hunk 6]
```diff
@@ -1693,3 +1694,32 @@ def test_positive_info_by_name(self):
...
+        for org in org_names:
+            Org.delete({'name': org})
+        if org_list:
```
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: This check is basically wrong. In case org_list is empty - no steps  will be performed and test will show us false positive. And, unfortunately, that's what really happened - `.remove('')` returned you `None`, `if org_list:` was skipped as org_list is None, test showed you 'PASSED' but it verified nothing.

Use assertions instead, `self.assertGreaterEqual(len(org_list), len(org_names))` in this case - this way you'll ensure we have enough orgs in output and fail the test otherwise. And, as a bonus, it will reduce unnecessary indentation :)

I'd also think of `self.assertIn` to ensure concrete orgs are present in output - as otherwise (lets say orgs with special characters appeared on the next page) the test won't really verify bugzilla's bug.

File: tests/foreman/cli/test_organization.py [hunk 7]
```diff
@@ -1693,3 +1694,32 @@ def test_positive_info_by_name(self):
...
+            for org_str in org_list:
+                width = 0
+                for char in org_str:
+                    eaw = unicodedata.east_asian_width(char)
+                    if eaw in ["Na", "N", "A", "H"]:  # Narrow, neutral,...
+                        width += 1
+                    else:  # Wide
+                        width += 2
```
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: Nitpick: Multiple checks and multiple loops made a huge indentation. How about a bit different style to reduce it: for each char add +1 width, and, if it's not in `["Na", "N", "A", "H"]`, add extra +1? Smth like that:
```python
for char in org_str:
    eaw = unicodedata.east_asian_width(char)
    width += 1
    if eaw not in ["Na", "N", "A", "H"]:
        width += 1
```

Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: Personally i'd use different format instead:
```python
for org_str in org_list:
    width = sum(
        1 if unicodedata.east_asian_width(char) in ["Na", "N", "A", "H"]
        else 2
        for char in org_str
    )
    self.assertEqual(len(org_list[0]), width)
```

**Comment by chbrown13** [comment]
Thanks all, changes have been pushed

File: tests/foreman/cli/test_organization.py [hunk 8]
```diff
@@ -1693,3 +1694,29 @@ def test_positive_info_by_name(self):
...
+            make_org({'name': org})
+        org_list = [line for line in Org.list(output_format='table') if line]
+        self.assertGreaterEqual(len(org_list), len(org_names))
+        for org_str in org_list:
+            width = 0
+            for char in org_str:
+                width = sum(
+                    1 if unicodedata.east_asian_width(char)
+                    in ["Na", "N", "A", "H"]
+                    else 2 for char in org_str
```
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: This `sum()` is what's gonna calculate `width` of `org_str` for you, you don't need `width = 0`and `for char in org_str:` anymore

Comment (chbrown13) [inline tests/foreman/cli/test_organization.py]: Fixed

Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: Not exactly :) The comprehension inside `sum()` contains a loop already which will navigate through each character in line. This way you don't need extra loop `for char in org_str:` outside, it basically breaks all logic which currently sounds like `for each char in line calculate sum of all characters inside that char ` :)

Comment (chbrown13) [inline tests/foreman/cli/test_organization.py]: Oops, sorry my bad. Removed

**Comment by abalakh** [comment]
@chbrown13 test results, please? Make sure to provide test results each time you change test logic, it's both preventing mistakes in test logic and proving your changes are valid.

**Comment by chbrown13** [comment]
Latest test results:

[root@localhost robottelo-fork]# pytest -vv tests/foreman/cli/test_organization.py -k 'test_positive_multibyte_latin1_org_names'
==================================================================== test session starts =====================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.7, py-1.4.34, pluggy-0.4.0 -- /usr/bin/python
cachedir: .cache
rootdir: /home/chris/Documents/robottelo-fork, inifile:
plugins: cov-2.5.1, mock-1.6.0, xdist-1.17.1, services-1.2.1
collected 70 items 
2017-06-29 10:50:40 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/cli/test_organization.py::OrganizationTestCase::test_positive_multibyte_latin1_org_names PASSED

==================================================================== 69 tests deselected =====================================================================
========================================================== 1 passed, 69 deselected in 85.80 seconds ==========================================================

**Comment by chbrown13** [comment]
@abalakh Is just the test result enough or is the entire verbose test output preferred?

**Comment by abalakh** [comment]
@chbrown13 test results is enough, test output is too big for a comment and basically contains sensitive information (like internal hostnames)

File: tests/foreman/cli/test_organization.py [hunk 9]
```diff
@@ -1693,3 +1694,27 @@ def test_positive_info_by_name(self):
...
+        """Hammer Multibyte and Latin-1 Org names break list pagination
+
+        :id: 4fa0afe7-6d0a-4c3e-a0fc-4ecb95c50fc9
+
+        :BZ: 1418412
+
+        :expectedresults: Multibyte and latin1 names need to be
+            displayed with consistent spacing
+        """
+        org_names = valid_org_names_list()
```
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: `valid_org_names_list()` is decorated with `@filtered_datapoint` which means it returns only 1 element if `run_one_datapoint` is set to `true` in `robottelo.properties`.
It may work ok for you locally (if `run_one_datapoint=false`), but `run_one_datapoint` is set to `true` on our CI, so instead of entire list you'll actually get only 1 random item (which may be something different from latin1/utf8).
In addition, according to test name and BZ description issue is not about alpha/alphanumeric names, so instead of reusing `valid_org_names_list()` which won't work anyway, consider forming your own list with cjk/latin1/utf8 names.

File: tests/foreman/cli/test_organization.py [hunk 10]
```diff
@@ -1693,3 +1694,27 @@ def test_positive_info_by_name(self):
...
+        for org in org_names:
+            make_org({'name': org})
+        org_list = [line for line in Org.list(output_format='table') if line]
```
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: What if Org.list() hasn't returned our latin1 names at all (because of a bug or pagination)? Imo it's worth asserting `org_names` elements are present inside strings of `org_list`.

Comment (chbrown13) [inline tests/foreman/cli/test_organization.py]: Ok, how should we handle the case where the name is not in the list? Just fail the test?

Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: Yep, just use assertions, in case assertion won't pass it will fail the test.

Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: Quick example off the top of my head:
```python
for name in org_names:
    self.assertTrue(any(name in line for line in org_list))
```

Comment (chbrown13) [inline tests/foreman/cli/test_organization.py]: Ok cool, thanks. I kept running into an error and realized when long org names are cut off with "...". Just fixed it and pushed.

**Comment by chbrown13** [comment]
Test results:

[root@localhost robottelo-fork]# pytest -vv tests/foreman/cli/test_organization.py -k 'test_positive_multibyte_latin1_org_names'
==================================================================================================================================== test session starts ====================================================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.7, py-1.4.34, pluggy-0.4.0 -- /usr/bin/python
cachedir: .cache
rootdir: /home/chris/Documents/robottelo-fork, inifile:
plugins: cov-2.5.1, mock-1.6.0, xdist-1.17.1, services-1.2.1
collected 70 items 
2017-06-29 12:13:59 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/cli/test_organization.py::OrganizationTestCase::test_positive_multibyte_latin1_org_names PASSED

==================================================================================================================================== 69 tests deselected ====================================================================================================================================
========================================================================================================================= 1 passed, 69 deselected in 40.14 seconds ==========================================================================================================================

File: tests/foreman/cli/test_organization.py (hunk 10 above)
Comment (abalakh) [inline tests/foreman/cli/test_organization.py]: What's the max name length which is not shortened? I mean, have you tried 50 chars explicitly? `random.randint(1, 50)` could return you 10 chars and you could miss the fact it will actually fail for 50 chars 'cause boundary value is, let's say, 30 chars. Better to double check as otherwise test can fail randomly :)

**Comment by chbrown13** [comment]
Makes sense, 30 will definitely work. Pushed

On Thu, Jun 29, 2017 at 12:18 PM, Andrii Balakhtar <notifications@github.com

**Comment by sghai** [comment]
@oshtaier : Pending on you for final comment and merge :)
